File: DataStructure/Stack/Task2.py
def isOperator(c):
    if c == '=' or c == '-' or c == '*' or c == '/' or c == '^' or c == '(' or c == ')':
        return True
    return False

def preToInfix(prefix):

    i = len(prefix) - 1
    stack = []
    
    while i >= 0:
        if isOperator(prefix[i]):
            str = '(' + stack.pop() + prefix[i] + stack.pop() + ')'
            stack.append(str)
            i -= 1
        else:
            stack.append(prefix[i])
            i -= 1
    return stack.pop()

str = "*-A/BC-/AKL"
print(preToInfix(str))

File: DataStructure/Stack/Task2.cpp
#include <iostream>
#include <stack>

using namespace std;

//Kiểm tra toán tử
bool isOperator(char x)
{
    switch(x)
    {
        case '+': case '-' : case '*' : case '/':
            return true;
    }
    return false;
}

string preToInfix(string pre)
{
    int length = pre.size();
    stack<string> st;
    //duyệt các giá trị trong chuỗi từ phần tử cuối đến phần tử đầu
    for(int i = length -1 ; i >= 0; i--)
    {
        char c = pre[i];
        //Nếu c là một toán tử, ta lấy 2 giá trị đầu của stack và toán tử sẽ đặt ở giữa 2 giá trị đó
        if(isOperator(c) == true)
        {
            string op1 = st.top(); st.pop();
            string op2 = st.top(); st.pop();
            string temp = "(" + op1 + pre[i] + op2 + ")";
        //Đưa string temp vào stack trở lại
            st.push(temp);
        }
        else
        {
            st.push(string(1,pre[i]));  // string(1,) đối số 1 nghĩa là tạo một chuỗi rỗng mới
        }
    }

    return st.top();    // lúc này trong chuỗi chỉ còn phần tử chuỗi kết quả 
}

int main() 
{
    string pre_exp = "*-A/BC-/AKL";
    cout << "Infix : " << preToInfix(pre_exp);
    return 0;
}
